[Database] TAPP-121: Add tables and attributes according to new schema plan

michellemtchai opened the pull request:
Please see issue #121 for details.

freeatnet commented:
@jmzaleski I think a few extra commits made their way into this branch, jfyi

freeatnet commented:
Things we need before we are ready to deploy:
- [ ] Figure out how to compile our React frontend down to a static JS file (we should not be running `webpack-dev-server` in production)
- [ ] Remove installation of development & test libraries from production image
- [ ] Maybe: do not install nodejs stuff in production image (since all JS should come precompiled)
- [ ] Split `docker-compose.yml` into base `docker-compose.yml` + per-environment overrides, `docker-compose.dev.yml` and `docker-compose.production.yml` (see below) and provide `COMPOSE_FILE` in `.env` files
- [ ] Confirm DB setup is aligned with departmental practices

Things that should be extracted into overrides:
* Development-only services (e.g., `webpack-dev-server` should only be defined in `docker-compose.dev.yml`)
* Commands: put the sensible default command, put development versions of them in development override
* Mounting of code into the container (only in development)
* Production config, like non-3000 port for Rails, reverse-proxies, HTTPS

Environment variables configurable from `.env` file should be in the base docker-compose file.

simpslandyy commented:
@michellemtchai @jmzaleski - I'll be modifying this a bit more so don't merge yet!

michellemtchai committed:
feat(db): remove attributes from applicants
remove the following attributes from applicants:
address
commentary
dept
year_in_program
is_full_time
is_grad_student
program
dept_fields

michellemtchai committed:
feat(db): add reporting_tags table
create reporting_tags table with attributes:
name:string

create reporting_tags model in rails

michellemtchai committed:
feat(db): create wage_chunk table
create wage_chunk table/model in rails containing the attributes:
hours:float
rate:float
start_date:datetime
end_date:datetime

michellemtchai committed:
feat(db): add has_many/belongs_to for reporting_tag and wage_chunk
add has_many :reporting_tags to wage_chunk
add belongs_to :wage_chunk to reporting_tag

michellemtchai committed:
feat(db): add has_many/belongs_to relation to position and reporting_tag
add has_many :reporting_tags to position
add belongs_to :position to reporting_tag

michellemtchai committed:
feat((db): remove attributes from positions
The following attributes were removed from positions table:
course_name
current_enrolment
duties
qualifications
hours
cap_enrolment
num_waitlisted
openings
start_date
end_date

michellemtchai committed:
fix(db): match remove_attributes_from_applicants to class name
fix issue with remove_attributes_from_applicants to class name as RemoveAttributesFromApplicants

michellemtchai committed:
feat(db): add new columns to position table
Add the following attributes to the position table:position_code
position_title
est_hours_per_assignment
est_start_date
est_end_date
position_type

michellemtchai committed:
feat(db): remove columns from sessions table
Remove the following attributes from sessions:
year
semester
pay

michellemtchai committed:
feat(db): add new columns to sessions table
The following attributes were added to sessions table:
name:string
rate1:float
rate2:float

michellemtchai committed:
feat(db): create position_template table/model
Create a position_template table with the following attributes:
position_type:string
offer_template:string

Creates a cooresponding position_template model in Rails

michellemtchai committed:
feat(db): add has_many/belongs_to to session and position_template model
add has_many :postion_templates to session model
add belongs_to :session to position_template model

michellemtchai committed:
feat(db): remove columns from assignments
Remove the following attributes from assignments table
applicant_id
position_id
hours
start_date
end_date
status

michellemtchai committed:
feat(db): add new columns to assignments
Add the following attributes to assignments table:
contract_start:datetime
contract_end:datetime
note:text
offer_override_pdf:string

michellemtchai committed:
feat(db): create preference_level table/model
Create preference_level table with the attributes:
preference_level:integer

Create the corresponding preference_level model

michellemtchai committed:
fix(db): fix preference_level model class name
rename create_preference_level.rb model/class to preference_model.rb

michellemtchai committed:
feat(db): remove columns from offers table
Removed the following attributes from offers table:
end_date
hours
start_date
status

michellemtchai committed:
feat(db): add new columns to offers table
Added the following attributes to offers table:
offer_template:string
offer_override_pdf:string
first_name:string
last_name:string
email:string
position_code:string
postion_title:string
position_start_date:datetime
position_end_date:datetime
first_time_ta:bool
instructor_contact_desc:string
pay_period_desc:string
installments:integer
ta_coordinator_name:string
ta_coordinator_email:string
emailed_date:datetime
signature:string
accepted_date:datetime
rejected_date:datetime
withdrawn_date:datetime

michellemtchai committed:
feat(db): create application table/model
Create application table with attributes:
comments:text

Create application model

michellemtchai committed:
feat(db): create preference_level table/model
Create preference_level table with the following attributes:
position_id
preference_level_id
application_id

michellemtchai committed:
feat(db): create relation between assignment, applicant, position
added position_id and applicant_id to assignment table
added has_many :assignment to both applicant and position model

michellemtchai committed:
feat(db): create relation between application and session
added session_id to application table
added belongs_to :session to application model
added has_many :applications to session model

michellemtchai committed:
feat(db): create relation between reporting_tag, position & wage_chunk
added position_id and wage_chunk_id to reporting_tags table

michellemtchai committed:
feat(db): create relation between wage_chunk & assignment
added assignment_id to wage_chunks table
added belongs_to :assignment in wage_chunks model
added has_many :wage_chunks in assignments model

michellemtchai committed:
feat(db): create relation between position_preference, position, application & preference_level
added belongs_to :position, :application, :preference_level to position_preference model
added has_many :position_preferences to preference_level
added has_many :position_preferences to position model
added a through to application on position model
added has_many :position_preferences to application model
added a through to position on application model

michellemtchai committed:
feat(db): add unique index for applicant-position pair in assignments
added unique index for (applicant_id, position_id) in assignment table
added validation for unique index for (applicant_id, position_id) in assignment model

michellemtchai committed:
feat(db): add unique index for application-position pair in position_preferences
added unique index for (application_id, position_id) in position_preferences table
added validation for unique index for (application_id, position_id) in position_preferences model

michellemtchai committed:
feat(db): remove preference table, model, related model references & spec
dropped preferences table
removed preferences model file
removed preference.rb and preference_spec from spec folder
remove all references of preferences model in other model files

michellemtchai committed:
fix(db): rename create_preference_levels table to preference_level
rename the table so that model references will work

michellemtchai committed:
feat(db): add position_data_for_ads table/model
created position_data_for_ads table/model
updated schema through new migration

michellemtchai committed:
feat(db): add relation between position & position_data_for_ad
added position_id to position_data_for_ads table
added belongs_to :position in position_data_for_ads model
added has_one :position_data_for_ad in positions model

michellemtchai committed:
feat(db): add position_data_for_matchings table/model
created position_data_for_matchings table/model
updated schema through new migration

michellemtchai committed:
feat(db): add relation between position & position_data_for_matching
added position_id to position_data_for_matchings table
added belongs_to :position in position_data_for_matchings model
added has_one :position_data_for_ad in positions model

michellemtchai committed:
feat(db): set columns in applicants to not null
set the following columns in applicants to not null via migration:
first_name
last_name
email
student_number
utorid

michellemtchai committed:
feat(db): set columns in instructors to not null
set the following columns in instructors to not null via migration:
first_name
last_name
email
utorid

michellemtchai committed:
feat(db): remove validation for deleted columns in positions
deleted not-null validation for non-existing columns
deleted unique pair validation for non-existing columns
deleted constraint validation for non-existing columns

michellemtchai committed:
feat(db): remove validation for deleted columns in sessions
deleted not-null validation for non-existing columns
deleted unique pair validation for non-existing columns
deleted constraint validation for non-existing columns
deleted enum declaration for non-existing columns

michellemtchai committed:
feat(db): set columns in users to not null
set the following columns in users to not null via migration:
utorid

George334456 commented on api/app/models/position.rb:
```diff
@@ -13,11 +13,10 @@ class Position < ApplicationRecord
+  has_one :position_data_for_matching
 
 
-  validates_presence_of :course_code, :openings, :session
```
Is there a reason we're deleting these rails validations? I don't think we have the database validations for these either.

michellemtchai replied:
I deleted these validation because the columns used for the validation no longer exist. The new schema plan got rid of them, so I wasn't if the exact same constraints should apply to which attribute.

George334456 commented on api/db/migrate/20190606151443_create_position_data_for_ads.rb:
```diff
@@ -0,0 +1,14 @@
+class CreatePositionDataForAds < ActiveRecord::Migration[5.2]
+  def change
+    create_table :position_data_for_ads do |t|
```
Did we ever come to a conclusion on whether or not we should be on one singular position table or multiple position tables?

michellemtchai replied:
@reidka mentioned in the meeting yesterday to separate position into multiple tables.

George334456 reviewed (commented):
Couple of comments, please respond to them.

Also, please start an issue for unit tests.

michellemtchai commented:
I've made a new issue #125 for the unit tests.

George334456 reviewed (approved):
Goooooooooooooooooooooooooooooooooooood stuff
